src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -5815,6 +5815,20 @@ private static void CombineExtensionMethodArguments(BoundExpression receiver, An
                 HashSet<DiagnosticInfo> useSiteDiagnostics = null;
                 OverloadResolution.MethodInvocationOverloadResolution(methodGroup.Methods, methodGroup.TypeArguments, actualArguments, overloadResolutionResult, ref useSiteDiagnostics, isMethodGroupConversion, allowRefOmittedArguments);
                 diagnostics.Add(expression, useSiteDiagnostics);
+
+                if (overloadResolutionResult.Succeeded)
+                {
+                    var receiverParameter = overloadResolutionResult.BestResult.Member.Parameters.FirstOrDefault();
+                    Debug.Assert(receiverParameter != null);
+
+                    if (receiverParameter.RefKind == RefKind.Ref)
+                    {
+                        // If this was a ref extension method, the receiver argument must be checked for L-value constraints.
+                        // This helper method will also replace the symbol with a BadExpressionSymbol if it was invalid.
+                        actualArguments.Arguments.SetItem(0, CheckValue(left, BindValueKind.RefOrOut, diagnostics));
+                    }
+                }
+
```

Comment:
It is safe to assume that valid extension methods will always have a first non-null parameter.
This check `CheckValue()` is done post-overload resolution, as a suitable candidate was found, and the caller `left` is known to be the first argument.
Note that this is not done for ref-readonly extension methods, as the compiler will create temps for r-values.